artifactory/commands/terraform/terraformpublish.go

```diff
@@ -0,0 +1,304 @@
+	if err := tpc.setRepoFromConfiguration(); err != nil {
+		return err
+	}
+	artDetails, err := tpc.serverDetails.CreateArtAuthConfig()
+	if err != nil {
+		return err
+	}
+	return utils.ValidateRepoExists(tpc.repo, artDetails)
+}
+
+func (tpc *TerraformPublishCommand) publish() error {
+	log.Debug("Deploying terraform module.")
+	success, failed, err := tpc.TerraformPublish()
+	if err != nil {
+		return err
+	}
+	tpc.result.SetSuccessCount(success)
+	tpc.result.SetFailCount(failed)
+	return nil
+}
+
+func ExtractTerraformPublishOptionsFromArgs(args []string) (namespace, provider, tag string, exclusions []string, err error) {
+	// Extract namespace information from the args.
+	flagIndex, valueIndex, namespace, err := coreutils.FindFlag("--namespace", args)
+	if err != nil {
+		return
+	}
+	coreutils.RemoveFlagFromCommand(&args, flagIndex, valueIndex)
+	// Extract provider information from the args.
+	flagIndex, valueIndex, provider, err = coreutils.FindFlag("--provider", args)
+	if err != nil {
+		return
+	}
+	coreutils.RemoveFlagFromCommand(&args, flagIndex, valueIndex)
+	// Extract tag information from the args.
+	flagIndex, valueIndex, tag, err = coreutils.FindFlag("--tag", args)
+	if err != nil {
+		return
+	}
+	coreutils.RemoveFlagFromCommand(&args, flagIndex, valueIndex)
+	// Extract exclusions information from the args.
+	flagIndex, valueIndex, exclusionsString, err := coreutils.FindFlag("--exclusions", args)
+	for _, singleValue := range strings.Split(exclusionsString, ";") {
+		exclusions = append(exclusions, singleValue)
+	}
+	if err != nil {
+		return
+	}
+	coreutils.RemoveFlagFromCommand(&args, flagIndex, valueIndex)
+	if len(args) != 0 {
+		err = errorutils.CheckErrorf("Unknown flag:\"" + strings.Split(args[0], "=")[0] + "\". for a terraform publish command please provide --namespace, --provider and --tag.")
+	}
+	return
+}
+
+func (ts *TerraformPublishCommand) TerraformPublish() (int, int, error) {
+	var e error
+	uploadSummary := clientservicesutils.NewResult(cliutils.Threads)
+	producerConsumer := parallel.NewRunner(cliutils.Threads, 20000, false)
+	errorsQueue := clientutils.NewErrorsQueue(1)
+
+	ts.prepareTerraformPublishTasks(producerConsumer, errorsQueue, uploadSummary)
+	totalUploaded, totalFailed := ts.performTerraformPublishTasks(producerConsumer, uploadSummary)
+	e = errorsQueue.GetError()
+	if e != nil {
+		return 0, 0, e
+	}
+	return totalUploaded, totalFailed, nil
+}
+
+func (ts *TerraformPublishCommand) prepareTerraformPublishTasks(producer parallel.Runner, errorsQueue *clientutils.ErrorsQueue, uploadSummary *clientservicesutils.Result) {
+	go func() {
+		defer producer.Done()
+		toArchive := make(map[string]*services.ArchiveUploadData)
+		pwd, err := os.Getwd()
+		if err != nil {
+			log.Error(err)
+			errorsQueue.AddError(err)
+		}
+		// Walk and upload directories which contain '.tf' files.
+		err = filepath.WalkDir(pwd, func(path string, info fs.DirEntry, err error) error {
```

Comment:
I suggest having a unit test, that checks that functionality of collecting directories with .tf files. To allow for this test, we may need to slightly refactor and isolate this walk operation. 

Also, if I understand the code correctly, the code works as follows -

1. Look for directory for upload
2. Stop
2. Upload the directory
3. Continue looking for the next directory

This is really inefficient, because of the "Stop" on step #2 above. 
I may have missed something while reading the code, so please let me know if I did. I just see the "walk" and the "upload" happening inside the same go routine.